### PR TITLE
Update embedding op signature

### PR DESCRIPTION
**Pull request by bmon581**
- Change embedding op signature which should be a DefaultDispatch operation. This is called both by C++ and Python APIs. This reduces code duplication between the two.
- This allows devices like MLC which needs to override Embedding with norm to be able to do so while keeping the current behavior.

### CODE REVIEW

File: torch/testing/_internal/jit_metaprogramming_utils.py [hunk 1]
```diff
@@ -106,7 +106,7 @@
     ('linear', (S, S), ((M, S),), '', (True, ['aten::linear'])),
     ('linear', (S, S), ((M, S), (M,)), 'addmm', (True, ['aten::linear'])),
     ('bilinear', (S, S, S), ((S, S, M), torch.zeros(M, S, M),),),
-    ('embedding', torch.tensor([[1, 2, 4, 5], [4, 3, 2, 5]]), (torch.rand(6, 3), ), '', (True,)),
+    ('embedding', torch.tensor([[1, 2, 4, 5], [4, 3, 2, 5]]), (torch.rand(6, 3), ), '',),
```
Comment (albanD) [inline torch/testing/_internal/jit_metaprogramming_utils.py:109]: Why do you need to change this?

**Review by albanD — COMMENTED** [review]
Thanks for sending this PR.

Can you clarify why this is a new operator and not replacing the existing embedding one (by adding optional arguments)?

And if we plan to keep this as a new operator, this is missing, test, doc and special note in the doc about why we have two embeddings now.

**Comment by belmhal** [comment]
> Thanks for sending this PR.
> 
> Can you clarify why this is a new operator and not replacing the existing embedding one (by adding optional arguments)?
> 
> And if we plan to keep this as a new operator, this is missing, test, doc and special note in the doc about why we have two embeddings now.

That is a good question. I was trying to keep the embedding layer in case that other backends were depending on its signature to override the op. (e.g. XLA). If we don't mind me changing the op signature, I can just remove embedding_with_norm and change the signature of embedding itself.

**Comment by albanD** [comment]
I think you want to update the existing embedding signature here.
We don't want to have these two functions that do the same thing I think.

**Comment by belmhal** [comment]
> I think you want to update the existing embedding signature here.
> We don't want to have these two functions that do the same thing I think.

They do almost the same thing. One does the normalization and one does not. That said, I completely agree with you. Let me update the PR to just update the existing signature. Thank you for the feedback! :)

**Comment by belmhal** [comment]
> > I think you want to update the existing embedding signature here.
> > We don't want to have these two functions that do the same thing I think.
> 
> They do almost the same thing. One does the normalization and one does not. That said, I completely agree with you. Let me update the PR to just update the existing signature. Thank you for the feedback! :)

updated to have embedding layer signature changed.

**Review by albanD — COMMENTED** [review]
That looks good!
I was about to ask for updating the doc of `torch.embedding` but there doesn't seem to be any doc for it so... 

Some small details in the the code but looks good otherwise.
Looks like you need to fix a couple jit tests where the signatures are hardcoded though.

File: aten/src/ATen/native/Embedding.cpp [hunk 2]
```diff
@@ -11,23 +12,47 @@
...
+
+Tensor embedding(
+    const Tensor& weight,
+    const Tensor& indices,
+    int64_t padding_idx,
+    bool scale_grad_by_freq,
+    bool sparse,
+    c10::optional<double> max_norm,
+    double norm_type) {
+  auto input_ = indices;
```
[the remark this answers is not in the document]
Comment (belmhal) [reply aten/src/ATen/native/Embedding.cpp:32]: That is a good suggestion! Thank you Alban, will update!

File: aten/src/ATen/native/Embedding.cpp [hunk 3]
```diff
@@ -11,23 +12,47 @@
 
 namespace at { namespace native {
 
-Tensor embedding(const Tensor & weight, const Tensor & indices,
-                 int64_t padding_idx, bool scale_grad_by_freq, bool sparse) {
+inline void _no_grad_embedding_renorm_(
```
[the remark this answers is not in the document]
Comment (belmhal) [reply aten/src/ATen/native/Embedding.cpp:15]: Yes, I noticed that. Beyond that what I noticed is that when the _no_grad_embedding_renorm_ and the embedding are part of the same operation, we no longer throw an exception on the test_embedding_renorm_grad_error. I am assuming that this expected behavior, after all we cannot backprop from the input to the embedding_renorm_ op. Hence, the user should receive a warning. Does that mean that the correct path would be the separate ops that are called by a default dispatch op like we had before? Or is there a simpler workaround in this instance?

**Comment by belmhal** [comment]
> That looks good!
> I was about to ask for updating the doc of `torch.embedding` but there doesn't seem to be any doc for it so...
> 
> Some small details in the the code but looks good otherwise.
> Looks like you need to fix a couple jit tests where the signatures are hardcoded though.

I updated the code as discussed offline. Please let me know if I missed something.
Hopefully the changes for the hardcoded JIT traces are now fixed.
I added some comments regarding the _embedding op, but please let me know if I missed something. Thank you again for reviewing this!

**Comment by belmhal** [comment]
@albanD I am having a hard time reproing the issues with the jit signatures locally. Do you have any idea of what could be different in the CI/CD pipeline vs. locally?

File: torch/csrc/jit/passes/shape_analysis.cpp [hunk 4]
```diff
@@ -1701,7 +1701,7 @@ class ShapePropagator {
       }
     } else if (
         node->matches(
-            "aten::embedding(Tensor weight, Tensor indices, int padding_idx, bool scale_grad_by_freq, bool sparse) -> Tensor")) {
+            "aten::embedding(Tensor weight, Tensor indices, int? padding_idx=None, bool scale_grad_by_freq=False, bool sparse=False, float? max_norm=None, float norm_type=2) -> Tensor")) {
```
Comment (albanD) [inline torch/csrc/jit/passes/shape_analysis.cpp:1704]: The jit test seems to still be failing.
Maybe changing this to detect `_embedding` would be simpler?

**Review by albanD — COMMENTED** [review]
The code logic lgtm
We just need to fix the jit test and this is good to go!

**Comment by belmhal** [comment]
@smessmer 

```
Mar 25 07:13:53 The PR is introducing backward incompatible changes to the operator library. Please contact PyTorch team to confirm whether this change is wanted or not. 
Mar 25 07:13:53 
Mar 25 07:13:53 Broken ops: [
Mar 25 07:13:53 	aten::embedding(Tensor weight, Tensor indices, int padding_idx=-1, bool scale_grad_by_freq=False, bool sparse=False) -> (Tensor)
Mar 25 07:13:53 ]
```

This is intended. Can I proceed with the changes or do I need to change something?

**Comment by albanD** [comment]
If you break the signature on purpose, you can add an entry to that list: https://github.com/pytorch/pytorch/blob/dfc7fa03e5d33f909b9d7853dd001086f5d782a0/test/backward_compatibility/check_backward_compatibility.py#L28
